diff --git a/frameworks/av/camera/CameraParameters.cpp b/frameworks/av/camera/CameraParameters.cpp
index 09e0e93..6fca89d 100644
--- a/frameworks/av/camera/CameraParameters.cpp
+++ b/frameworks/av/camera/CameraParameters.cpp
@@ -268,6 +268,9 @@ const char CameraParameters::WHITE_BALANCE_SHADE[] = "shade";
 #ifdef OPPO_CAMERA_HARDWARE
 const char CameraParameters::WHITE_BALANCE_MANUAL_CCT[] = "manual-cct";
 #endif
+#ifdef NUBIA_CAMERA_HARDWARE
+const char CameraParameters::WHITE_BALANCE_MANUAL_CCTE[] = "manual-ccte";
+#endif
 
 // Values for effect settings.
 const char CameraParameters::EFFECT_NONE[] = "none";
diff --git a/frameworks/av/include/camera/CameraParameters.h b/frameworks/av/include/camera/CameraParameters.h
index 4a5ec8f..2949483 100644
--- a/frameworks/av/include/camera/CameraParameters.h
+++ b/frameworks/av/include/camera/CameraParameters.h
@@ -776,6 +776,10 @@ public:
     static const char WHITE_BALANCE_MANUAL_CCT[];
 #endif
 
+#ifdef NUBIA_CAMERA_HARDWARE
+    static const char WHITE_BALANCE_MANUAL_CCTE[];
+#endif
+
     // Values for effect settings.
     static const char EFFECT_NONE[];
     static const char EFFECT_MONO[];
