diff --git a/vendor/cm/config/telephony.mk b/vendor/cm/config/telephony.mk
index 1c0c4a6..a279355 100644
--- a/vendor/cm/config/telephony.mk
+++ b/vendor/cm/config/telephony.mk
@@ -1,10 +1,10 @@
 # World APN list
-PRODUCT_COPY_FILES += \
-    vendor/cm/prebuilt/common/etc/apns-conf.xml:system/etc/apns-conf.xml
+#PRODUCT_COPY_FILES += \
+#    vendor/cm/prebuilt/common/etc/apns-conf.xml:system/etc/apns-conf.xml
 
 # World SPN overrides list
-PRODUCT_COPY_FILES += \
-    vendor/cm/prebuilt/common/etc/spn-conf.xml:system/etc/spn-conf.xml
+#PRODUCT_COPY_FILES += \
+#    vendor/cm/prebuilt/common/etc/spn-conf.xml:system/etc/spn-conf.xml
 
 # Selective SPN list for operator number who has the problem.
 PRODUCT_COPY_FILES += \
