diff --git a/build/core/Makefile b/build/core/Makefile
index 0e815fd..991465d 100644
--- a/build/core/Makefile
+++ b/build/core/Makefile
@@ -1465,7 +1465,6 @@ $(INTERNAL_OTA_PACKAGE_TARGET): $(BUILT_TARGET_FILES_PACKAGE) $(DISTTOOLS)
 	   -k $(KEY_CERT_PAIR) \
 	   --backup=$(backuptool) \
 	   --override_device=$(override_device) \
-	   --override_prop=$(override_prop) \
 	   $(BUILT_TARGET_FILES_PACKAGE) $@
 
 CM_TARGET_PACKAGE := $(PRODUCT_OUT)/rr-$(CM_VERSION).zip
diff --git a/build/core/combo/arch/arm/armv7-a-neon.mk b/build/core/combo/arch/arm/armv7-a-neon.mk
index 398d0e8..a4380f1 100644
--- a/build/core/combo/arch/arm/armv7-a-neon.mk
+++ b/build/core/combo/arch/arm/armv7-a-neon.mk
@@ -6,6 +6,13 @@ ARCH_ARM_HAVE_VFP               := true
 ARCH_ARM_HAVE_VFP_D32           := true
 ARCH_ARM_HAVE_NEON              := true
 
+TARGET_USE_KRAIT_BIONIC_OPTIMIZATION := true
+TARGET_USE_KRAIT_PLD_SET := true
+TARGET_KRAIT_BIONIC_PLDOFFS := 10
+TARGET_KRAIT_BIONIC_PLDTHRESH := 10
+TARGET_KRAIT_BIONIC_BBTHRESH := 64
+TARGET_KRAIT_BIONIC_PLDSIZE := 64
+
 ifeq ($(strip $(TARGET_CPU_VARIANT)), cortex-a15)
 	arch_variant_cflags := -mcpu=cortex-a15
 else
@@ -22,7 +29,7 @@ ifeq ($(strip $(TARGET_CPU_VARIANT)),cortex-a5)
 	arch_variant_cflags := -mcpu=cortex-a5
 else
 ifeq ($(strip $(TARGET_CPU_VARIANT)),krait)
-	arch_variant_cflags := -mcpu=cortex-a9
+	arch_variant_cflags := -mcpu=cortex-a15
 else
 ifeq ($(strip $(TARGET_CPU_VARIANT)),scorpion)
 	arch_variant_cflags := -mcpu=cortex-a8
@@ -38,7 +45,7 @@ endif
 
 arch_variant_cflags += \
     -mfloat-abi=softfp \
-    -mfpu=neon
+    -mfpu=neon-vfpv4
 
 arch_variant_ldflags := \
 	-Wl,--fix-cortex-a8
