
project system/vold/
diff --git a/system/vold/Ext4.cpp b/system/vold/Ext4.cpp
index 7981f7e..7f23cef 100644
--- a/system/vold/Ext4.cpp
+++ b/system/vold/Ext4.cpp
@@ -67,6 +67,7 @@ int Ext4::doMount(const char *fsPath, const char *mountPoint, bool ro, bool remo
         if (data[0])
             strlcat(data, ",", sizeof(data));
         strlcat(data, "context=u:object_r:sdcard_external:s0", sizeof(data));
+        strlcat(data, ",discard,noauto_da_alloc,errors=panic,data=ordered", sizeof(data));
     }
 
     rc = mount(fsPath, mountPoint, "ext4", flags, data);
