
project bionic/
diff --git a/bionic/libc/arch-arm/bionic/libgcc_compat.c b/bionic/libc/arch-arm/bionic/libgcc_compat.c
index f57dfb0..e3b4439 100644
--- a/bionic/libc/arch-arm/bionic/libgcc_compat.c
+++ b/bionic/libc/arch-arm/bionic/libgcc_compat.c
@@ -119,6 +119,7 @@
     XX(__aeabi_i2d)          \
     XX(__aeabi_i2f)          \
     XX(__aeabi_idiv)         \
+    XX(__aeabi_idivmod)      \
     XX(__aeabi_l2d)          \
     XX(__aeabi_l2f)          \
     XX(__aeabi_lasr)         \
@@ -129,6 +130,8 @@
     XX(__aeabi_ui2f)         \
     XX(__aeabi_ul2d)         \
     XX(__aeabi_ul2f)         \
+    XX(__aeabi_uidiv)        \
+    XX(__aeabi_uidivmod)     \
     XX(__cmpdf2)             \
     XX(__divdf3)             \
     XX(__divsf3)             \
@@ -152,6 +155,8 @@
     XX(__muldi3)             \
     XX(__mulsf3)             \
     XX(__nedf2)              \
+    XX(__popcountsi2)        \
+    XX(__popcount_tab)       \
     XX(__subdf3)             \
     XX(__subsf3)             \
     XX(__truncdfsf2)         \
